File: lib/emltpl_to_oft/mapi.py
from __future__ import annotations

import struct
import time
from pathlib import Path

from .cfb import CLSID_OFT, CFBWriter, _python_time_to_filetime

PT_INT32 = 0x0003
PT_BOOLEAN = 0x000B
PT_STRING = 0x001F
PT_BINARY = 0x0102
PT_SYSTIME = 0x0040

PROP_RW = 0x00000006

MSGFLAG_UNSENT = 0x00000008
MSGFLAG_HASATTACH = 0x00000010

DEFAULT_INTERNET_CODEPAGE = 65001


def _utf16le(text: str) -> bytes:
    return text.encode("utf-16-le")


def _stream_name(prop_id: int, prop_type: int) -> str:
    return f"__substg1.0_{prop_id:04X}{prop_type:04X}"


def _short_filename(name: str) -> str:
    path = Path(name)
    return f"{path.stem[:8]}{path.suffix[:4]}"


class PropertyStreamBuilder:
    """Build a __properties_version1.0 stream and its value streams."""

    def __init__(self, is_top_level: bool = False):
        self._is_top_level = is_top_level
        self._fixed: list[bytes] = []
        self._streams: dict[str, bytes | memoryview] = {}
        self._n_recipients = 0
        self._n_attachments = 0

    def set_counts(self, recipients: int, attachments: int) -> None:
        self._n_recipients = recipients
        self._n_attachments = attachments

    def add_int32(self, prop_id: int, value: int) -> None:
        entry = struct.pack("<HHI", PT_INT32, prop_id, PROP_RW)
        entry += struct.pack("<I", value & 0xFFFFFFFF) + b"\x00" * 4
        self._fixed.append(entry)

    def add_bool(self, prop_id: int, value: bool) -> None:
        entry = struct.pack("<HHI", PT_BOOLEAN, prop_id, PROP_RW)
        entry += struct.pack("<I", 1 if value else 0) + b"\x00" * 4
        self._fixed.append(entry)

    def add_time(self, prop_id: int, filetime: int) -> None:
        entry = struct.pack("<HHI", PT_SYSTIME, prop_id, PROP_RW)
        entry += struct.pack("<Q", filetime)
        self._fixed.append(entry)

    def add_string(self, prop_id: int, value: str) -> None:
        data = _utf16le(value)
        name = _stream_name(prop_id, PT_STRING)
        self._streams[name] = data
        entry = struct.pack("<HHI", PT_STRING, prop_id, PROP_RW)
        entry += struct.pack("<I", len(data)) + b"\x00" * 4
        self._fixed.append(entry)

    def add_binary(self, prop_id: int, data: bytes | memoryview) -> None:
        name = _stream_name(prop_id, PT_BINARY)
        self._streams[name] = data
        entry = struct.pack("<HHI", PT_BINARY, prop_id, PROP_RW)
        entry += struct.pack("<I", len(data)) + b"\x00" * 4
        self._fixed.append(entry)

    def build_props_stream(self) -> bytes:
        if self._is_top_level:
            header = struct.pack(
                "<QIIII",
                0,
                self._n_recipients,
                self._n_attachments,
                self._n_recipients,
                self._n_attachments,
            )
            header += b"\x00" * 8
        else:
            header = b"\x00" * 8
        return header + b"".join(self._fixed)

    @property
    def value_streams(self) -> dict[str, bytes | memoryview]:
        return dict(self._streams)


class OFTBuilder:
    """Build an OFT file from decoded message components."""

    def __init__(self) -> None:
        self.subject: str = ""
        self.body_text: str = ""
        self.body_html: bytes = b""
        self.internet_codepage = DEFAULT_INTERNET_CODEPAGE
        self.attachments: list[dict] = []

    def build(self, path: str | Path) -> None:
        cfb = CFBWriter()

        root = cfb.add_root(clsid=CLSID_OFT)

        nameid = cfb.add_storage(root, "__nameid_version1.0")
        cfb.add_stream(nameid, "__substg1.0_00020102", b"")
        cfb.add_stream(nameid, "__substg1.0_00030102", b"")
        cfb.add_stream(nameid, "__substg1.0_00040102", b"")

        props = PropertyStreamBuilder(is_top_level=True)
        n_attach = len(self.attachments)
        props.set_counts(recipients=0, attachments=n_attach)

        now_ft = _python_time_to_filetime(time.time())
        msg_flags = MSGFLAG_UNSENT | (MSGFLAG_HASATTACH if n_attach > 0 else 0)

        props.add_string(0x001A, "IPM.Note")
        props.add_string(0x0037, self.subject)
        props.add_string(0x003D, "")
        props.add_string(0x0070, self.subject)
        props.add_string(0x0E1D, self.subject)
        props.add_int32(0x0E07, msg_flags)
        props.add_int32(0x340D, 0x00040E79)
        props.add_int32(0x3FDE, self.internet_codepage)
        props.add_int32(0x3FF1, 0x0409)
        props.add_time(0x3007, now_ft)
        props.add_time(0x3008, now_ft)

        if self.body_text:
            props.add_string(0x1000, self.body_text)
        if self.body_html:
            props.add_binary(0x1013, self.body_html)

        cfb.add_stream(root, "__properties_version1.0", props.build_props_stream())
        for stream_name, stream_data in props.value_streams.items():
            cfb.add_stream(root, stream_name, stream_data)

        for attachment_index, attachment in enumerate(self.attachments):
            attachment_storage = cfb.add_storage(
                root, f"__attach_version1.0_#{attachment_index:08X}"
            )
            props = PropertyStreamBuilder(is_top_level=False)

            props.add_int32(0x3705, 0x00000001)
            props.add_int32(0x370B, 0xFFFFFFFF)
            props.add_int32(0x0E20, len(attachment["data"]))
            props.add_int32(0x0FFE, 0x00000007)

            filename = attachment.get("filename", f"attachment_{attachment_index}")
            props.add_string(0x3707, filename)
            props.add_string(0x3704, _short_filename(filename))
            props.add_string(0x3001, filename)

            mime_type = attachment.get("mime_type", "application/octet-stream")
            props.add_string(0x370E, mime_type)

            extension = Path(filename).suffix
            if extension:
                props.add_string(0x3703, extension)

            content_id = attachment.get("content_id")
            if content_id:
                props.add_string(0x3712, content_id)

            props.add_binary(0x3701, attachment["data"])
            props.add_binary(0x3702, b"")

            cfb.add_stream(
                attachment_storage,
                "__properties_version1.0",
                props.build_props_stream(),
            )
            for stream_name, stream_data in props.value_streams.items():
                cfb.add_stream(attachment_storage, stream_name, stream_data)

        cfb.save(path)

File: packages/converter/src/mapi.ts
import { basename, extname } from 'node:path'
import { CFBWriter, CLSID_OFT, unixMillisecondsToFiletime } from './cfb'

export const PT_INT32 = 0x0003
const PT_BOOLEAN = 0x000b
const PT_STRING = 0x001f
const PT_BINARY = 0x0102
const PT_SYSTIME = 0x0040

const PROP_RW = 0x00000006

const MSGFLAG_UNSENT = 0x00000008
const MSGFLAG_HASATTACH = 0x00000010

export const DEFAULT_INTERNET_CODEPAGE = 65001

type ByteBuffer = Buffer<ArrayBufferLike>

type AttachmentInput = {
  contentId: string | null
  data: ByteBuffer
  disposition: 'attachment' | 'inline' | null
  filename: string
  mimeType: string
}

const utf16le = (text: string) => Buffer.from(text, 'utf16le')

const streamName = (propId: number, propType: number) =>
  `__substg1.0_${propId.toString(16).toUpperCase().padStart(4, '0')}${propType
    .toString(16)
    .toUpperCase()
    .padStart(4, '0')}`

const shortFilename = (name: string) => {
  const extension = extname(name).slice(0, 4)
  const stem = basename(name, extension).slice(0, 8)
  return `${stem}${extension}`
}

class PropertyStreamBuilder {
  readonly #fixed: Buffer[] = []
  readonly #streams = new Map<string, Buffer>()
  #attachmentCount = 0
  readonly #isTopLevel: boolean
  #recipientCount = 0

  constructor(isTopLevel = false) {
    this.#isTopLevel = isTopLevel
  }

  setCounts(recipients: number, attachments: number) {
    this.#recipientCount = recipients
    this.#attachmentCount = attachments
  }

  addInt32(propId: number, value: number) {
    const entry = Buffer.alloc(16)
    entry.writeUInt16LE(PT_INT32, 0)
    entry.writeUInt16LE(propId, 2)
    entry.writeUInt32LE(PROP_RW, 4)
    entry.writeUInt32LE(value >>> 0, 8)
    this.#fixed.push(entry)
  }

  addBool(propId: number, value: boolean) {
    const entry = Buffer.alloc(16)
    entry.writeUInt16LE(PT_BOOLEAN, 0)
    entry.writeUInt16LE(propId, 2)
    entry.writeUInt32LE(PROP_RW, 4)
    entry.writeUInt32LE(value ? 1 : 0, 8)
    this.#fixed.push(entry)
  }

  addTime(propId: number, filetime: bigint) {
    const entry = Buffer.alloc(16)
    entry.writeUInt16LE(PT_SYSTIME, 0)
    entry.writeUInt16LE(propId, 2)
    entry.writeUInt32LE(PROP_RW, 4)
    entry.writeBigUInt64LE(filetime, 8)
    this.#fixed.push(entry)
  }

  addString(propId: number, value: string) {
    const data = utf16le(value)
    this.#streams.set(streamName(propId, PT_STRING), data)

    const entry = Buffer.alloc(16)
    entry.writeUInt16LE(PT_STRING, 0)
    entry.writeUInt16LE(propId, 2)
    entry.writeUInt32LE(PROP_RW, 4)
    entry.writeUInt32LE(data.length, 8)
    this.#fixed.push(entry)
  }

  addBinary(propId: number, value: Buffer) {
    this.#streams.set(streamName(propId, PT_BINARY), Buffer.from(value))

    const entry = Buffer.alloc(16)
    entry.writeUInt16LE(PT_BINARY, 0)
    entry.writeUInt16LE(propId, 2)
    entry.writeUInt32LE(PROP_RW, 4)
    entry.writeUInt32LE(value.length, 8)
    this.#fixed.push(entry)
  }

  buildPropsStream() {
    if (this.#isTopLevel) {
      const header = Buffer.alloc(32)
      header.writeBigUInt64LE(0n, 0)
      header.writeUInt32LE(this.#recipientCount >>> 0, 8)
      header.writeUInt32LE(this.#attachmentCount >>> 0, 12)
      header.writeUInt32LE(this.#recipientCount >>> 0, 16)
      header.writeUInt32LE(this.#attachmentCount >>> 0, 20)
      return Buffer.concat([header, ...this.#fixed])
    }

    return Buffer.concat([Buffer.alloc(8), ...this.#fixed])
  }

  get valueStreams() {
    return new Map(this.#streams)
  }
}

export class OFTBuilder {
  attachments: AttachmentInput[] = []
  bodyHtml: ByteBuffer = Buffer.alloc(0)
  bodyText = ''
  internetCodepage = DEFAULT_INTERNET_CODEPAGE
  subject = ''

  toBuffer() {
    const cfb = new CFBWriter()
    const root = cfb.addRoot(CLSID_OFT)

    const nameId = cfb.addStorage(root, '__nameid_version1.0')
    cfb.addStream(nameId, '__substg1.0_00020102', Buffer.alloc(0))
    cfb.addStream(nameId, '__substg1.0_00030102', Buffer.alloc(0))
    cfb.addStream(nameId, '__substg1.0_00040102', Buffer.alloc(0))

    const topLevelProps = new PropertyStreamBuilder(true)
    topLevelProps.setCounts(0, this.attachments.length)

    const now = unixMillisecondsToFiletime(Date.now())
    const messageFlags = MSGFLAG_UNSENT | (this.attachments.length > 0 ? MSGFLAG_HASATTACH : 0)

    topLevelProps.addString(0x001a, 'IPM.Note')
    topLevelProps.addString(0x0037, this.subject)
    topLevelProps.addString(0x003d, '')
    topLevelProps.addString(0x0070, this.subject)
    topLevelProps.addString(0x0e1d, this.subject)
    topLevelProps.addInt32(0x0e07, messageFlags)
    topLevelProps.addInt32(0x340d, 0x00040e79)
    topLevelProps.addInt32(0x3fde, this.internetCodepage)
    topLevelProps.addInt32(0x3ff1, 0x0409)
    topLevelProps.addTime(0x3007, now)
    topLevelProps.addTime(0x3008, now)

    if (this.bodyText.length > 0) {
      topLevelProps.addString(0x1000, this.bodyText)
    }

    if (this.bodyHtml.length > 0) {
      topLevelProps.addBinary(0x1013, this.bodyHtml)
    }

    cfb.addStream(root, '__properties_version1.0', topLevelProps.buildPropsStream())
    for (const [name, value] of topLevelProps.valueStreams) {
      cfb.addStream(root, name, value)
    }

    this.attachments.forEach((attachment, index) => {
      const attachmentStorage = cfb.addStorage(
        root,
        `__attach_version1.0_#${index.toString(16).toUpperCase().padStart(8, '0')}`,
      )

      const props = new PropertyStreamBuilder(false)
      props.addInt32(0x3705, 0x00000001)
      props.addInt32(0x370b, 0xffffffff)
      props.addInt32(0x0e20, attachment.data.length)
      props.addInt32(0x0ffe, 0x00000007)
      props.addString(0x3707, attachment.filename)
      props.addString(0x3704, shortFilename(attachment.filename))
      props.addString(0x3001, attachment.filename)
      props.addString(0x370e, attachment.mimeType)

      const extension = extname(attachment.filename)
      if (extension.length > 0) {
        props.addString(0x3703, extension)
      }

      if (attachment.contentId) {
        props.addString(0x3712, attachment.contentId)
      }

      props.addBinary(0x3701, attachment.data)
      props.addBinary(0x3702, Buffer.alloc(0))

      cfb.addStream(attachmentStorage, '__properties_version1.0', props.buildPropsStream())
      for (const [name, value] of props.valueStreams) {
        cfb.addStream(attachmentStorage, name, value)
      }
    })

    return cfb.toBuffer()
  }
}
